in_toto/keylib.go

```diff
@@ -2,220 +2,330 @@ package in_toto
+	data, _ := pem.Decode(pemBytes)
+	if data == nil {
+		return ErrNoPEMBLock
 	}
 
-	// 64 hexadecimal digits => 32 bytes for the private portion of the key
-	if len(keyObj.KeyVal.Private) != 64 {
-		return keyObj, fmt.Errorf("the private field on this key is malformed")
+	// Try to load private key, if this fails try to load
+	// key as public key
+	key, err := ParseKey(data.Bytes)
+	if err != nil {
+		return err
 	}
 
-	return keyObj, nil
+	// Use type switch to identify the key format
+	switch key.(type) {
+	case *rsa.PublicKey:
+		if err := k.SetKeyComponents(pemBytes, []byte{}, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case *rsa.PrivateKey:
+		// Note: We store the public key as PKCS8 key here, although the private key get's stored as PKCS1 key
+		// This behavior is consistent to the securesystemslib
+		pubKeyBytes, err := x509.MarshalPKIXPublicKey(key.(*rsa.PrivateKey).Public())
+		if err != nil {
+			return err
+		}
+		if err := k.SetKeyComponents(pubKeyBytes, pemBytes, "rsa", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PublicKey:
+		if err := k.SetKeyComponents(key.(ed25519.PublicKey), []byte{}, "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	case ed25519.PrivateKey:
+		pubKeyBytes := key.(ed25519.PrivateKey).Public()
+		if err := k.SetKeyComponents(pubKeyBytes.(ed25519.PublicKey), key.(ed25519.PrivateKey), "ed25519", scheme, keyIdHashAlgorithms); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("%w: %T", ErrUnsupportedKeyType, key)
+	}
+	return nil
 }
 
 /*
-GenerateEd25519Signature creates an ed25519 signature using the key and the
-signable buffer provided. It returns an error if the underlying signing library
-fails.
-*/
-func GenerateEd25519Signature(signable []byte, key Key) (Signature, error) {
+GenerateSignature will automatically detect the key type and sign the signable data
+with the provided key. If everything goes right GenerateSignature will return
+a for the key valid signature and err=nil. If something goes wrong it will
+return an not initialized signature and an error. Possible errors are:
 
-	var signature Signature
+	* ErrNoPEMBlock
+	* ErrUnsupportedKeyType
 
-	seed, err := hex.DecodeString(key.KeyVal.Private)
-	if err != nil {
-		return signature, err
+*/
+func GenerateSignature(signable []byte, key Key) (Signature, error) {
+	var signature Signature
+	var signatureBuffer []byte
+	// The following switch block is needed for keeping interoperability
+	// with the securesystemslib and the python implementation
+	// in which we are storing RSA keys in PEM format, but ed25519 keys hex encoded.
+	switch key.KeyType {
+	case "rsa":
+		keyReader := strings.NewReader(key.KeyVal.Private)
+		pemBytes, err := ioutil.ReadAll(keyReader)
+		if err != nil {
+			return signature, err
+		}
+		// TODO: There could be more key data in _, which we silently ignore here.
+		// Should we handle it / fail / say something about it?
```

Comment:
what type of "key data" are we talking about? could we add at least an example in the comments? It'd help anybody who wants to work this in the future to scope what they are getting into.

Reply:
According to the  pem [documentation](https://golang.org/pkg/encoding/pem/#Decode), we are dropping the rest of the key data here. The rest is everything that couldn't be parsed as pem block.

Reply:
I have added more documentation for this and removed the "TODO".